src/hotspot/share/opto/castnode.hpp

```diff
@@ -32,30 +32,95 @@
 //------------------------------ConstraintCastNode-----------------------------
 // cast to a different range
 class ConstraintCastNode: public TypeNode {
-public:
-  enum DependencyType {
-    RegularDependency, // if cast doesn't improve input type, cast can be removed
-    StrongDependency,  // leave cast in even if _type doesn't improve input type, can be replaced by stricter dominating cast if one exist
-    UnconditionalDependency // leave cast in unconditionally
+protected:
+  // Cast nodes are subject to a few optimizations:
+  //
+  // 1- if the type carried by the Cast doesn't narrow the type of its input, the cast can be replaced by its input.
+  // Similarly, if a dominating Cast with the same input and a narrower type constraint is found, it can replace the
+  // current cast.
+  //
+  // 2- if the condition that the Cast is control dependent is hoisted, the Cast is hoisted as well
+  //
+  // 1- and 2- are not always applied depending on what constraint are applied to the Cast: there are cases where 1-
+  // and 2- apply, where neither 1- nor 2- apply and where one or the other apply. This class abstract away these
+  // details.
+  class DependencyType {
+  public:
+    DependencyType(bool depends_on_test, bool narrows_type, const char* desc)
+      : _depends_only_on_test(depends_on_test),
+        _narrows_type(narrows_type),
+        _desc(desc) {
+    }
+    NONCOPYABLE(DependencyType);
+
+    bool depends_only_on_test() const {
+      return _depends_only_on_test;
+    }
+
+    bool narrows_type() const {
+      return _narrows_type;
+    }
+    void dump_on(outputStream *st) const {
+      st->print("%s", _desc);
+    }
+
+    uint hash() const {
+      return (_depends_only_on_test ? 1 : 0) + (_narrows_type ? 2 : 0);
+    }
+
+    bool cmp(const DependencyType& other) const {
+      return _depends_only_on_test == other._depends_only_on_test && _narrows_type == other._narrows_type;
+    }
+
+    const DependencyType& widen_type_dependency() const {
+      if (_depends_only_on_test) {
+        return WidenTypeDependency;
+      }
+      return UnconditionalDependency;
+    }
+
+    const DependencyType& pinned_dependency() const {
+      if (_narrows_type) {
+        return StrongDependency;
+      }
+      return UnconditionalDependency;
+    }
+
+  private:
+    const bool _depends_only_on_test; // Does this Cast depends on its control input or is it pinned?
+    const bool _narrows_type; // Does this Cast narrows the type i.e. if input type is narrower can it be removed?
```

Comment:
I think it would be good to have a really strong definition of these two, because everything else depends on it.

I would recommend to either use `depends_only_on_test` as the "primary" word here, or else `pinned`. But then try to consistently use the chosen one everywhere. Just to avoid confusion with these near synonymes.

It may also be helpful to have an example for each of the 4 combinations, just as an illustration of your definitions.